### PR TITLE
Steps 1 & 2

### PR DESCRIPTION
**Pull request by CaptainJNS**

### CODE REVIEW

File: app/controllers/books_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class BooksController < ApplicationController
+  def index
+    @sorts = ['Newest first', 'Title: A-Z ↑', 'Title: Z-A ↓', 'Price: to UP ↑', 'Price: to DOWN ↓']
```
Comment (dzemlianoi-double) [inline app/controllers/books_controller.rb]: move to query object or service object

File: app/models/book.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  has_and_belongs_to_many :authors
+  has_and_belongs_to_many :categories
+  has_many_attached :images
+
+  validates :title, :description, :price, :year, :quantity, presence: true
+
+  validates :price, numericality: { greater_or_equal_to: 0, less_than_or_equal_to: 100_000 }
+  validates :quantity, numericality: { only_integer: true, greater_or_equal_to: 0 }
+
+  scope :most_recent,      -> { order('created_at DESC') }
```
Comment (dzemlianoi-double) [inline app/models/book.rb]: move to query of sorting

File: app/views/books/_books.html.haml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            %a.thumb-hover-link{href: book_path(book)}
+              %i.fa.fa-eye.thumb-icon{'aria-hidden': "true"}
+            %a.thumb-hover-link{href: "#"}
+              %i.fa.fa-shopping-cart.thumb-icon
+    .general-thumb-info
+      %p.title= link_to book.title, book_path(book)
+      %p.font-16.in-gold-500= '€'+book.price.to_s
+      %span.divider-sm
+      %p.lead.small
+        = book.authors.map{ |author| author.name }.join(', ')
```
Comment (dzemlianoi-double) [inline app/views/books/_books.html.haml]: decorator

File: app/views/books/_books.html.haml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+          = image_tag('no_cover.jpg', class: 'img-shadow general-thumbnail-img')
+        .thumb-hover
+          .general-thumb-link-wrap
+            %a.thumb-hover-link{href: book_path(book)}
+              %i.fa.fa-eye.thumb-icon{'aria-hidden': "true"}
+            %a.thumb-hover-link{href: "#"}
+              %i.fa.fa-shopping-cart.thumb-icon
+    .general-thumb-info
+      %p.title= link_to book.title, book_path(book)
+      %p.font-16.in-gold-500= '€'+book.price.to_s
```
Comment (dzemlianoi-double) [inline app/views/books/_books.html.haml]: number_to_currency

File: app/views/books/_filter.html.haml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+
+%ul.list-inline.pt-10.mb-25.mr-240
+  %li.mr-35
+    %a.filter-link{href: books_path}
+      All
+      %span.badge.general-badge= Book.all.size
```
Comment (dzemlianoi-double) [inline app/views/books/_filter.html.haml]: count

File: app/views/books/_filter.html.haml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  %li.mr-35
+    %a.filter-link{href: "#"}
+      - Category.all.each do |category|
```
Comment (dzemlianoi-double) [inline app/views/books/_filter.html.haml]: find_each

Comment (dzemlianoi-double) [inline app/views/books/_filter.html.haml]: move to helper or service

Comment (dzemlianoi-double) [inline app/views/books/_filter.html.haml]: Categoru.includes(:books)

Comment (dzemlianoi-double) [reply app/views/books/_filter.html.haml]: use Bullet

File: app/views/books/_mobile_filter.html.haml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+%a.dropdown-toggle.lead.small{href: "#", 'data-toggle': "dropdown", role: "button", 'aria-haspopup': "true", 'aria-expanded': "false"}
+  All
+  %i.fa.fa-angle-down.dropdown-icon.pull-right
+%ul.dropdown-menu
+  %li
+    %a{href: "#"} All
+  - Category.all.each do |category|
```
Comment (dzemlianoi-double) [inline app/views/books/_mobile_filter.html.haml]: use pluck and find_each

File: app/views/books/_mobile_sort.html.haml [hunk 8]
```diff
@@ -0,0 +1,12 @@
+%a.dropdown-toggle.lead.small{href: "#", 'data-toggle': "dropdown", role: "button", 'aria-haspopup': "true", 'aria-expanded': "false"}
+  Newest first
```
Comment (dzemlianoi-double) [inline app/views/books/_mobile_sort.html.haml]: i18n and take it from query object

File: app/views/books/_sorting.html.haml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+.dropdowns.pull-right
+  .dropdown.width-240
+    %a.dropdown-toggle.lead.small{href: "#", 'data-toggle': "dropdown", role: "button", 'aria-haspopup': "true", 'aria-expanded': "false"}
+      = @sorts[params[:sort_by].to_i]
+      %i.fa.fa-angle-down.dropdown-icon.pull-right
+    %ul.dropdown-menu
+      - @sorts.each_with_index do |sort, index|
+        %li
+          %a{href: books_path(category: params[:category], sort_by: index)}
```
Comment (dzemlianoi-double) [inline app/views/books/_sorting.html.haml]: = link_to sort, books_path(category: params[:category], sort_by: index)

File: app/views/books/show.html.haml [hunk 10]
```diff
@@ -0,0 +1,63 @@
+= content_for(:title, @book.title + ' | ')
+%main.container
+  = link_to :back, class: 'general-back-link' do
+    %i.fa.fa-long-arrow-left.mr-15
+    Back to results
+  .row
+    .col-sm-6
+      .product-gallery
+        - if @book.images.any?
```
Comment (dzemlianoi-double) [inline app/views/books/show.html.haml]: move to partial

File: app/views/books/show.html.haml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            - @book.images.each do |image|
+              %a.img-link
+                = image_tag(image.variant(resize: "150x150>"), class: 'img-responsive')
+        - else
+          = image_tag('no_cover.jpg', class: 'img-shadow general-thumbnail-img')
+
+    .col-sm-6
+      %h1.mt-res-0= @book.title
+      %p.lead.small
+        = @book.authors.map{ |author| author.name }.join(', ')
```
Comment (dzemlianoi-double) [inline app/views/books/show.html.haml]: decorator

File: app/views/books/show.html.haml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+            %input.form-control.quantity-input{type: "text", value: "1"}
+            %a.input-link{href: "#"}
+              %i.fa.fa-plus.line-height-40
+        .col-xs-7.col-md-5
+          %button.btn.btn-default.pull-right.general-position Add to Cart
+      .general-info-wrap.divider-lg.description
+        %p Description
+        %p.lead.small.line-height-2
+          = truncate(@book.description, length: 250)
+          - if @book.description.size > 250
```
Comment (dzemlianoi-double) [inline app/views/books/show.html.haml]: magic numbers

File: app/views/books/show.html.haml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+            = link_to 'Read more', @book, class: 'in-gold-500 ml-10', remote: true
+      .general-info-wrap.divider-lg
+        .clearfix
+          %p.pull-left Year of publication
+          %p.lead.small.pull-right= @book.year
+      .general-info-wrap.divider-lg
+        .clearfix
+          %p.pull-left Category
+          %p.lead.small.pull-right= @book.categories.map{ |category| category.name }.join(', ')
```
Comment (dzemlianoi-double) [inline app/views/books/show.html.haml]: decorator

File: app/views/layouts/application.html.haml [hunk 14]
```diff
@@ -0,0 +1,22 @@
+!!!
+%html
+  %head
+    %meta{charset: "UTF-8"}
+    %meta{name: "viewport", content: "width=device-width"}
+    %title= yield(:title) + 'Bookstore'
+    %link{href: "https://fonts.googleapis.com/css?family=Open+Sans", rel: "stylesheet"}
+    = csrf_meta_tags
+    = stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload'
+    -# %link{href: "../css/style.css", rel: "stylesheet"}
```
Comment (dzemlianoi-double) [inline app/views/layouts/application.html.haml]: remove it please

File: config/routes.rb [hunk 15]
```diff
@@ -1,3 +1,9 @@
 Rails.application.routes.draw do
-  # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+  # get 'books/index'
+  # get 'books/show'
+  resources :books
+
+  get 'home/index'
```
Comment (dzemlianoi-double) [inline config/routes.rb]: remove

File: db/seeds.rb [hunk 16]
```diff
@@ -1,7 +1,45 @@
-# This file should contain all the record creation needed to seed the database with its default values.
-# The data can then be loaded with the rails db:seed command (or created alongside the database with db:setup).
-#
-# Examples:
-#
-#   movies = Movie.create([{ name: 'Star Wars' }, { name: 'Lord of the Rings' }])
-#   Character.create(name: 'Luke', movie: movies.first)
+CATEGORIES = ['Mobile development', 'Photo', 'Web design', 'Web development'].freeze
```
Comment (dzemlianoi-double) [inline db/seeds.rb]: Seed bank

File: spec/controllers/books_controller_spec.rb [hunk 17]
```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+
+RSpec.describe BooksController, type: :controller do
+  describe 'GET #show' do
+    let(:book) { FactoryBot.create(:book) }
```
Comment (dzemlianoi-double) [inline spec/controllers/books_controller_spec.rb]: create(:book)

Comment (dzemlianoi-double) [reply spec/controllers/books_controller_spec.rb]: RSpec.configure do |config|
  config.include FactoryBot::Syntax::Methods
end

File: spec/factories/books.rb [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+  factory :book do
+    title { FFaker::Book.unique.title }
+    description { FFaker::Book.description }
+    price {rand(5.0..500.00)}
+    quantity {rand(1..100)}
+    dimensions {"#{rand(DIMENSION).round(1)}\"x#{rand(DIMENSION).round(1)}\"x#{rand(DIMENSION).round(1)}\""}
+    year {rand(1666..2019)}
+    materials {MATERIALS.sample(rand(1..3)).join(', ')}
+    categories {Array.new(1) { create(:category) } }
+    authors { Array.new(2) { create(:author) } }
```
Comment (dzemlianoi-double) [inline spec/factories/books.rb]: create_list(:author, 2)

File: spec/factories/categories.rb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+CATEGORIES = ['Mobile development', 'Photo', 'Web design', 'Web development'].freeze
```
Comment (dzemlianoi-double) [inline spec/factories/categories.rb]: move to FFaker

File: spec/rails_helper.rb [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+require 'spec_helper'
+ENV['RAILS_ENV'] ||= 'test'
+require File.expand_path('../../config/environment', __FILE__)
+# Prevent database truncation if the environment is production
+abort("The Rails environment is running in production mode!") if Rails.env.production?
+require 'rspec/rails'
+# Add additional requires below this line. Rails is not loaded until this point!
+require 'simplecov'
+
+SimpleCov.start do
```
Comment (dzemlianoi-double) [inline spec/rails_helper.rb]: move to support

File: app/views/shared/_menu.html.haml [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+      %li
+        %a{href: "#"} Mobile development
+      %li
+        %a{href: "#"} Photo
+      %li
+        %a{href: "#"} Web design
+      %li.divider{role: "separator"}
+      %li
+        %a{href: "#"} Log out
+    -# - if user_signed_in?
```
Comment (vatnajokull) [inline app/views/shared/_menu.html.haml]: please delete

File: config/database.yml [hunk 22]
```diff
@@ -5,21 +5,25 @@
...
+  username: jason
+  password: password
+  host: localhost
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   timeout: 5000
 
 development:
   <<: *default
-  database: db/development.sqlite3
+  database: development
```
Comment (vatnajokull) [inline config/database.yml]: Please provide uniq name of database name, for example, `bookstore_development`

File: config/sidekiq.yml [hunk 23]
```diff
@@ -0,0 +1,6 @@
+---
```
Comment (vatnajokull) [inline config/sidekiq.yml]: Why do you need this file?

File: db/migrate/20190817104429_create_join_table_book_category.rb [hunk 24]
```diff
@@ -0,0 +1,8 @@
+class CreateJoinTableBookCategory < ActiveRecord::Migration[5.2]
+  def change
+    create_join_table :books, :categories do |t|
+      # t.index [:book_id, :category_id]
```
Comment (vatnajokull) [inline db/migrate/20190817104429_create_join_table_book_category.rb]: ??

File: db/schema.rb [hunk 25]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  create_table 'authors', force: :cascade do |t|
+    t.string 'first_name'
+    t.string 'last_name'
+    t.datetime 'created_at', null: false
+    t.datetime 'updated_at', null: false
+  end
+
+  create_table 'authors_books', id: false, force: :cascade do |t|
```
Comment (vatnajokull) [inline db/schema.rb]: Do you really need this table?

File: spec/models/author_book_spec.rb [hunk 26]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
+
+RSpec.describe AuthorBook, type: :model do
+  # pending "add some examples to (or delete) #{__FILE__}"
```
Comment (vatnajokull) [inline spec/models/author_book_spec.rb]: any tests?

File: app/views/books/_description.html.haml [hunk 27]
```diff
@@ -0,0 +1,4 @@
+.general-info-wrap.divider-lg
+  %p Description
```
[the remark this answers is not in the document]
Comment (CaptainJNS) [reply app/views/books/_description.html.haml]: I'll do it later, at the end of developing

[the remark this answers is not in the document]
Comment (vatnajokull) [reply app/views/books/_description.html.haml]: I don't think it's a good idea. Could you please start using i18n from the start of the project?

File: app/views/books/_filter.html.haml [hunk 28]
```diff
@@ -2,11 +2,11 @@
...
-      %span.badge.general-badge= Book.count
+      = t('shop.all')
+      %span.badge.general-badge= @book_count
   %li.mr-35
     %a.filter-link{href: "#"}
-      - Category.includes(:books).find_each do |category|
+      - @categories.each do |category|
         %li.mr-35
           .filter-link
             = link_to category.name, books_path(category: category.id), class: 'collapse-link'
```
Comment (vatnajokull) [inline app/views/books/_filter.html.haml]: Seems like this link follow me to books#show with the id of a selected category. Is it expected behavior?

File: app/views/books/_mobile_filter.html.haml [hunk 29]
```diff
@@ -1,9 +1,9 @@
...
 %ul.dropdown-menu
   %li
-    %a{href: "#"} All
-  - Category.all.each do |category|
+    %a{href: "#"}= t('shop.all')
+  - @categories.each do |category|
     %li
-      %a{href: "#"}= category.name
\ No newline at end of file
+      = link_to category.name, books_path(category: category.id)
```
Comment (vatnajokull) [inline app/views/books/_mobile_filter.html.haml]: The same as above

File: app/controllers/books_controller.rb [hunk 30]
```diff
@@ -0,0 +1,14 @@
+class BooksController < ApplicationController
+  def index
+    @book_count = Book.count
+    @categories = Category.select(:id, :name)
+    params[:sort_by] ||= 'created_at DESC'
```
Comment (vatnajokull) [inline app/controllers/books_controller.rb]: Please delete this

File: app/queries/default_books_query.rb [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def latest_books(relation)
+    return relation unless @params.include?(:latest_books)
+
+    relation.last(Constants::LATEST_BOOK_COUNT)
+  end
+
+  def filter(relation)
+    return relation unless Category.exists?(@params[:category])
```
Comment (vatnajokull) [inline app/queries/default_books_query.rb]: You can just check that value by this key in params is present.

File: app/queries/default_books_query.rb [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+
+    Category.find(@params[:category]).books
```
Comment (vatnajokull) [inline app/queries/default_books_query.rb]: Maybe it would be better to use `relation` for finding all books from some category?

**Pull request closed without merging**